File: system_monitor/collector/registry.py
from __future__ import annotations

from typing import Callable

from ..config_loader import SensorConfig
from .base import Sensor
from .gpio import Dht22Sensor
from .remote import HttpJsonSensor, MqttSensor
from .system import (
    CpuPercentSensor,
    DiskUsageSensor,
    MemoryPercentSensor,
    NetworkBytesSensor,
)
from .temperature import GpuTemperatureSensor, TemperatureSensor

SensorFactory = Callable[[SensorConfig], Sensor]

_REGISTRY: dict[str, SensorFactory] = {
    "system.cpu_percent": CpuPercentSensor,
    "system.memory_percent": MemoryPercentSensor,
    "system.disk_usage": DiskUsageSensor,
    "system.network_bytes": NetworkBytesSensor,
    "system.temperature": TemperatureSensor,
    "system.gpu_temperature": GpuTemperatureSensor,
    "gpio.dht22": Dht22Sensor,
    "remote.http_json": HttpJsonSensor,
    "remote.mqtt": MqttSensor,
}


def get_registered_types() -> list[str]:
    return sorted(_REGISTRY.keys())


def create_sensor(config: SensorConfig) -> Sensor:
    factory = _REGISTRY.get(config.type)
    if factory is None:
        raise ValueError(f"Неизвестный тип датчика: {config.type}")
    return factory(config)

File: internal/collector/registry.go
package collector

import (
	"fmt"

	"github.com/pagbest154-cmd/system-monitor/internal/config"
)

type factory func(config.SensorConfig) Sensor

var registry = map[string]factory{
	"system.cpu_percent":      func(c config.SensorConfig) Sensor { return cpuPercentSensor{baseSensor{c}} },
	"system.memory_percent":   func(c config.SensorConfig) Sensor { return memoryPercentSensor{baseSensor{c}} },
	"system.disk_usage":       func(c config.SensorConfig) Sensor { return diskUsageSensor{baseSensor{c}} },
	"system.network_bytes":    func(c config.SensorConfig) Sensor { return networkBytesSensor{baseSensor{c}} },
	"system.temperature":      func(c config.SensorConfig) Sensor { return temperatureSensor{baseSensor{c}} },
	"system.gpu_temperature":  func(c config.SensorConfig) Sensor { return gpuTemperatureSensor{baseSensor{c}} },
	"gpio.dht22":              func(c config.SensorConfig) Sensor { return dht22Sensor{baseSensor{c}} },
	"remote.http_json":        func(c config.SensorConfig) Sensor { return httpJsonSensor{baseSensor{c}} },
	"remote.mqtt":             func(c config.SensorConfig) Sensor { return mqttSensor{baseSensor{c}} },
}

func GetRegisteredTypes() []string {
	types := make([]string, 0, len(registry))
	for t := range registry {
		types = append(types, t)
	}
	sortStrings(types)
	return types
}

func sortStrings(s []string) {
	for i := 0; i < len(s); i++ {
		for j := i + 1; j < len(s); j++ {
			if s[j] < s[i] {
				s[i], s[j] = s[j], s[i]
			}
		}
	}
}

func CreateSensor(cfg config.SensorConfig) (Sensor, error) {
	f, ok := registry[cfg.Type]
	if !ok {
		return nil, fmt.Errorf("неизвестный тип датчика: %s", cfg.Type)
	}
	return f(cfg), nil
}
